Wasm component breakage between v0.20 and v0.21

In my research project, I am compiling several components (which implement different compression methods) and then run them with wasmtime (and a browser WebAssembly runtime). Running one of these components recently broke, all the others continue to work. After some debugging, I found that changing the wit-bindgen version back to 0.20 doesn't cause the error, but 0.21 (and 0.22) both have the issue.

I have attached both versions of the component (apologies for the size - it includes the SZ3 compressor which is the only one that breaks). For 0.21, I get the following error (in wasmtime) when invoking the codec-id function, which takes no inputs and returns a string:

```
RuntimeError: in fcbench:codec/codecs@0.1.0.codec-id: error while executing at wasm backtrace:
    0: 0x32943f - <unknown>!<wasm function 2566>
    1: 0x363007 - <unknown>!<wasm function 5920>
    2: 0x3617f4 - <unknown>!<wasm function 5527>
```

From debugging in the other backend, I think the error may occur during the cabi post-return function.

Apologies also for not having a minimal reproducing case - for some reason, all smaller compressor components continue to work, so there might be some spooky side effects at a distance at play. Thank you in advance for any insights you can provide!

[sz3-0.20.wasm](https://github.com/bytecodealliance/wit-bindgen/files/14591037/sz3-0.20.zip)
[sz3-0.21.wasm](https://github.com/bytecodealliance/wit-bindgen/files/14591039/sz3-0.21.zip)

> In the [released commit range](https://github.com/bytecodealliance/wit-bindgen/compare/v0.20.0...v0.21.0) what you're probably running into is #876. That being said I don't know why that would be causing issues here.
> 
> Are you sure that wit-bindgen is all that changed between those two builds?
> 
> In 0.20.wasm above I see:
> 
> ```webassembly
>     (func (;6433;) (type 1) (param i32 i32 i32 i32) (result i32)
>       local.get 0
>       local.get 1
>       local.get 2
>       local.get 3
>       call 6434
>     )
> ;; ...
>     (export "cabi_realloc" (func 6433))
> ```
> 
> This is what I would expect, that's just a small shim over another function.
> 
> In 0.21.wasm, however, I see:
> 
> ```webassembly
>     (func (;6459;) (type 1) (param i32 i32 i32 i32) (result i32)
>       call 13
>       local.get 0
>       local.get 1
>       local.get 2
>       local.get 3
>       call 6434
>       call 4716
>     )
> ;; ...
>     (export "cabi_realloc" (func 6459))
> ```
> 
> That doesn't look right, specifically `call 13` at the start and `call 4716` at the end.
> 
> IIRC that was something some part of the toolchain used to do where it injected constructors/destructors explicitly around exports. There was a comment in wit-bindgen saying after Rust 1.69 it was no longer necessary so I turned it off by default. In the components though I see that 1.76 was used to compile both, so now I'm a bit confused.
> 
> Can you try specifying `run_ctors_once_workaround: true,` in the 0.21 build and see if it fixes the issue? If it does then I would be confused. That changed in #868 which was part of the 0.20.0 release, so both builds should in theory not work then...

Thank you so much for looking into this!

> Are you sure that wit-bindgen is all that changed between those two builds?

Yes, my only change was changing the version in my Cargo.toml dependencies section.

> That doesn't look right, specifically `call 13` at the start and `call 4716` at the end.

That is indeed weird and interesting that it only causes an issue with this component but not the others. The Sz3 component is one of a few that also need to compile a C library that is wrapped by a *-sys Rust crate that I use, for which I'm using the [wasi-sdk](https://github.com/WebAssembly/wasi-sdk) v20.0, but the other components also with C parts still work ... intruiging.

> Can you try specifying `run_ctors_once_workaround: true,` in the 0.21 build and see if it fixes the issue? If it does then I would be confused. That changed in #868 which was part of the 0.20.0 release, so both builds should in theory not work then...

In v0.20, v0.21, and v0.22 that unfortunately gives this compilation error:

```rust
error[E0425]: cannot find function `bool_lift` in module `_rt`
    --> codecs/core-wasm/src/lib.rs:18:5
     |
  18 | /     wit_bindgen::generate!({
  19 | |         path: "../wit",
  20 | |         world: "fcbench-codec",
  21 | |         pub_export_macro: true,
  22 | |         run_ctors_once_workaround: true,
  23 | |     });
     | |______^ not found in `_rt`
     |
     = help: consider importing this function:
             wit_bindgen::rt::bool_lift
```

Is there some other option I need to specify?

--- a/crates/guest-rust/Cargo.toml
+++ b/crates/guest-rust/Cargo.toml
@@ -13,10 +13,10 @@ Used when compiling Rust programs to the component model.
 
 [dependencies]
 wit-bindgen-rust-macro = { path = "./macro", optional = true, version = "0.22.0" }
-wit-bindgen-rt = { path = "./rt", optional = true, version = "0.22.0" }
+wit-bindgen-rt = { path = "./rt", version = "0.22.0" }
 bitflags = { workspace = true }
 
 [features]
 default = ["macros", "realloc"]
 macros = ["dep:wit-bindgen-rust-macro"]
-realloc = ["dep:wit-bindgen-rt"]
+realloc = []

--- a/crates/guest-rust/macro/src/lib.rs
+++ b/crates/guest-rust/macro/src/lib.rs
@@ -101,8 +101,8 @@ impl Parse for Config {
                     Opt::TypeSectionSuffix(suffix) => {
                         opts.type_section_suffix = Some(suffix.value());
                     }
-                    Opt::RunCtorsOnceWorkaround(enable) => {
-                        opts.run_ctors_once_workaround = enable.value();
+                    Opt::DisableRunCtorsOnceWorkaround(enable) => {
+                        opts.disable_run_ctors_once_workaround = enable.value();
                     }
                     Opt::DefaultBindingsModule(enable) => {
                         opts.default_bindings_module = Some(enable.value());
@@ -227,7 +227,7 @@ mod kw {
     syn::custom_keyword!(additional_derives);
     syn::custom_keyword!(with);
     syn::custom_keyword!(type_section_suffix);
-    syn::custom_keyword!(run_ctors_once_workaround);
+    syn::custom_keyword!(disable_run_ctors_once_workaround);
     syn::custom_keyword!(default_bindings_module);
     syn::custom_keyword!(export_macro_name);
     syn::custom_keyword!(pub_export_macro);
@@ -276,7 +276,7 @@ enum Opt {
     AdditionalDerives(Vec<syn::Path>),
     With(HashMap<String, String>),
     TypeSectionSuffix(syn::LitStr),
-    RunCtorsOnceWorkaround(syn::LitBool),
+    DisableRunCtorsOnceWorkaround(syn::LitBool),
     DefaultBindingsModule(syn::LitStr),
     ExportMacroName(syn::LitStr),
     PubExportMacro(syn::LitBool),
@@ -382,10 +382,10 @@ impl Parse for Opt {
             input.parse::<kw::type_section_suffix>()?;
             input.parse::<Token![:]>()?;
             Ok(Opt::TypeSectionSuffix(input.parse()?))
-        } else if l.peek(kw::run_ctors_once_workaround) {
-            input.parse::<kw::run_ctors_once_workaround>()?;
+        } else if l.peek(kw::disable_run_ctors_once_workaround) {
+            input.parse::<kw::disable_run_ctors_once_workaround>()?;
             input.parse::<Token![:]>()?;
-            Ok(Opt::RunCtorsOnceWorkaround(input.parse()?))
+            Ok(Opt::DisableRunCtorsOnceWorkaround(input.parse()?))
         } else if l.peek(kw::default_bindings_module) {
             input.parse::<kw::default_bindings_module>()?;
             input.parse::<Token![:]>()?;

--- a/crates/guest-rust/rt/src/lib.rs
+++ b/crates/guest-rust/rt/src/lib.rs
@@ -79,3 +79,29 @@ pub unsafe fn cabi_realloc(
     }
     return ptr;
 }
+
+/// Provide a hook for generated export functions to run static constructors at
+/// most once.
+///
+/// wit-bindgen-rust generates a call to this function at the start of all
+/// component export functions. Importantly, it is not called as part of
+/// `cabi_realloc`, which is a *core* export func, but should not execute ctors.
+#[cfg(target_arch = "wasm32")]
+pub fn run_ctors_once() {
+    static mut RUN: bool = false;
+    unsafe {
+        if !RUN {
+            // This function is synthesized by `wasm-ld` to run all static
+            // constructors. wasm-ld will either provide an implementation
+            // of this symbol, or synthesize a wrapper around each
+            // exported function to (unconditionally) run ctors. By using
+            // this function, the linked module is opting into "manually"
+            // running ctors.
+            extern "C" {
+                fn __wasm_call_ctors();
+            }
+            __wasm_call_ctors();
+            RUN = true;
+        }
+    }
+}

--- a/crates/guest-rust/src/lib.rs
+++ b/crates/guest-rust/src/lib.rs
@@ -782,9 +782,9 @@
 ///     // support the standard library itself depending on this crate one day.
 ///     std_feature,
 ///
-///     // Force a workaround to be emitted for pre-Rust-1.69.0 modules to
-///     // ensure that libc ctors run only once.
-///     run_ctors_once_workaround: true,
+///     // Disable a workaround to force wasm constructors to be run only once
+///     // when exported functions are called.
+///     disable_run_ctors_once_workaround: false,
 /// });
 /// ```
 ///
@@ -803,6 +803,9 @@ pub mod examples;
 
 #[doc(hidden)]
 pub mod rt {
+    #[cfg(target_arch = "wasm32")]
+    pub use wit_bindgen_rt::run_ctors_once;
+
     pub fn maybe_link_cabi_realloc() {
         #[cfg(feature = "realloc")]
         wit_bindgen_rt::maybe_link_cabi_realloc();

--- a/crates/guest-rust/src/pre_wit_bindgen_0_20_0.rs
+++ b/crates/guest-rust/src/pre_wit_bindgen_0_20_0.rs
@@ -19,35 +19,6 @@ use core::sync::atomic::{AtomicU32, Ordering::Relaxed};
 
 pub use alloc_crate::{alloc, boxed, string, vec};
 
-/// Provide a hook for generated export functions to run static
-/// constructors at most once. wit-bindgen-rust generates a call to this
-/// function at the start of all component export functions. Importantly,
-/// it is not called as part of `cabi_realloc`, which is a *core* export
-/// func, but may not execute ctors, because the environment ctor in
-/// wasi-libc (before rust 1.69.0) calls an import func, which is not
-/// permitted by the Component Model when inside realloc.
-///
-/// We intend to remove this once rust 1.69.0 stabilizes.
-#[cfg(target_arch = "wasm32")]
-pub fn run_ctors_once() {
-    static mut RUN: bool = false;
-    unsafe {
-        if !RUN {
-            // This function is synthesized by `wasm-ld` to run all static
-            // constructors. wasm-ld will either provide an implementation
-            // of this symbol, or synthesize a wrapper around each
-            // exported function to (unconditionally) run ctors. By using
-            // this function, the linked module is opting into "manually"
-            // running ctors.
-            extern "C" {
-                fn __wasm_call_ctors();
-            }
-            __wasm_call_ctors();
-            RUN = true;
-        }
-    }
-}
-
 pub unsafe fn dealloc(ptr: i32, size: usize, align: usize) {
     if size == 0 {
         return;

--- a/crates/rust/src/interface.rs
+++ b/crates/rust/src/interface.rs
@@ -519,25 +519,19 @@ macro_rules! {macro_name} {{
         let params = self.print_export_sig(func);
         self.push_str(" {");
 
-        if self.gen.opts.run_ctors_once_workaround {
+        if !self.gen.opts.disable_run_ctors_once_workaround {
             let run_ctors_once = self.path_to_run_ctors_once();
+            // Before executing any other code, use this function to run all
+            // static constructors, if they have not yet been run. This is a
+            // hack required to work around wasi-libc ctors calling import
+            // functions to initialize the environment.
+            //
+            // See
+            // https://github.com/bytecodealliance/preview2-prototyping/issues/99
+            // for more details.
             uwrite!(
                 self.src,
-                "\
-                // Before executing any other code, use this function to run all static
-                // constructors, if they have not yet been run. This is a hack required
-                // to work around wasi-libc ctors calling import functions to initialize
-                // the environment.
-                //
-                // This functionality will be removed once rust 1.69.0 is stable, at which
-                // point wasi-libc will no longer have this behavior.
-                //
-                // See
-                // https://github.com/bytecodealliance/preview2-prototyping/issues/99
-                // for more details.
-                #[cfg(target_arch=\"wasm32\")]
-                {run_ctors_once}();
-",
+                "#[cfg(target_arch=\"wasm32\")]\n{run_ctors_once}();",
             );
         }
 
@@ -1917,7 +1911,7 @@ macro_rules! {macro_name} {{
     }
 
     fn path_to_run_ctors_once(&mut self) -> String {
-        self.path_from_runtime_module(RuntimeItem::RunCtorsOnce, "bool_lift")
+        self.path_from_runtime_module(RuntimeItem::RunCtorsOnce, "run_ctors_once")
     }
 
     pub fn path_to_vec(&mut self) -> String {

--- a/crates/rust/src/lib.rs
+++ b/crates/rust/src/lib.rs
@@ -164,10 +164,10 @@ pub struct Opts {
     #[cfg_attr(feature = "clap", arg(long))]
     pub type_section_suffix: Option<String>,
 
-    /// Apply a workaround required before Rust 1.69 to run wasm ctors only
-    /// once.
+    /// Disable a workaround used to prevent libc ctors/dtors from being invoked
+    /// too much.
     #[cfg_attr(feature = "clap", arg(long))]
-    pub run_ctors_once_workaround: bool,
+    pub disable_run_ctors_once_workaround: bool,
 
     /// Changes the default module used in the generated `export!` macro to
     /// something other than `self`.
@@ -413,36 +413,15 @@ pub unsafe fn bool_lift(val: u8) -> bool {
             }
 
             RuntimeItem::RunCtorsOnce => {
-                self.src.push_str(
+                let rt = self.runtime_path();
+                self.src.push_str(&format!(
                     r#"
-/// Provide a hook for generated export functions to run static
-/// constructors at most once. wit-bindgen-rust generates a call to this
-/// function at the start of all component export functions. Importantly,
-/// it is not called as part of `cabi_realloc`, which is a *core* export
-/// func, but may not execute ctors, because the environment ctor in
-/// wasi-libc (before rust 1.69.0) calls an import func, which is not
-/// permitted by the Component Model when inside realloc.
 #[cfg(target_arch = "wasm32")]
-pub fn run_ctors_once() {
-    static mut RUN: bool = false;
-    unsafe {
-        if !RUN {
-            // This function is synthesized by `wasm-ld` to run all static
-            // constructors. wasm-ld will either provide an implementation
-            // of this symbol, or synthesize a wrapper around each
-            // exported function to (unconditionally) run ctors. By using
-            // this function, the linked module is opting into "manually"
-            // running ctors.
-            extern "C" {
-                fn __wasm_call_ctors();
-            }
-            __wasm_call_ctors();
-            RUN = true;
-        }
-    }
-}
+pub fn run_ctors_once() {{
+    {rt}::run_ctors_once();
+}}
                     "#,
-                );
+                ));
             }
 
             RuntimeItem::AsI32 => {
@@ -868,8 +847,8 @@ impl WorldGenerator for RustWasm {
         if let Some(default) = &self.opts.default_bindings_module {
             uwriteln!(self.src, "//   * default-bindings-module: {default:?}");
         }
-        if self.opts.run_ctors_once_workaround {
-            uwriteln!(self.src, "//   * run-ctors-once-workaround");
+        if self.opts.disable_run_ctors_once_workaround {
+            uwriteln!(self.src, "//   * disable-run-ctors-once-workaround");
         }
         if let Some(s) = &self.opts.export_macro_name {
             uwriteln!(self.src, "//   * export-macro-name: {s}");

--- a/crates/rust/tests/codegen.rs
+++ b/crates/rust/tests/codegen.rs
@@ -86,7 +86,7 @@ mod run_ctors_once_workaround {
                 export apply-the-workaround: func();
             }
         ",
-        run_ctors_once_workaround: true,
+        disable_run_ctors_once_workaround: true,
         stubs,
     });
 }

